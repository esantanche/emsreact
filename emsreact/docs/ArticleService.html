<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ArticleService</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ArticleService</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ArticleService<span class="signature">()</span><span class="type-signature"></span></h2>
        
            <div class="class-description"><p>This service fetches articles from the backend.</p></div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="ArticleService"><span class="type-signature"></span>new ArticleService<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="services_article_ArticleService.js.html">services/article/ArticleService.js</a>, <a href="services_article_ArticleService.js.html#line10">line 10</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="fetch_articles"><span class="type-signature"></span>fetch_articles<span class="signature">(filters, callback_to_return_articles)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    <p>Function fetching articles.</p>
<p>There are three implemented filters.</p>
<p>The first one fetches a single article. It needs the node id the backend assigned to the article.</p>
<p>The second fetches all articles for a given topic. Topics can be &quot;How I work&quot;, &quot;The tools I use&quot;, etc.</p>
<p>The third fetches sticky articles only or all af them.</p>
<p>It's possible to combine the topic and sticky filters. Of course the nid filter (single article) has to be used alone.</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>filters</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"><p>Filters to use when fetching articles</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>callback_to_return_articles</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last"><p>Function to call to return the fetched articles</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="services_article_ArticleService.js.html">services/article/ArticleService.js</a>, <a href="services_article_ArticleService.js.html#line28">line 28</a>
    </li></ul></dd>
    

    

    

    
</dl>


















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="AlignedGridPane.html">AlignedGridPane</a></li><li><a href="App.html">App</a></li><li><a href="ArticleCard.html">ArticleCard</a></li><li><a href="ArticleScreen.html">ArticleScreen</a></li><li><a href="ArticleService.html">ArticleService</a></li><li><a href="BodyPane.html">BodyPane</a></li><li><a href="BottomGridPane.html">BottomGridPane</a></li><li><a href="ColouredCard.html">ColouredCard</a></li><li><a href="FooterScreen.html">FooterScreen</a></li><li><a href="FrontGridPane.html">FrontGridPane</a></li><li><a href="HeaderMenu.html">HeaderMenu</a></li><li><a href="HeaderScreen.html">HeaderScreen</a></li><li><a href="HomeButtonPane.html">HomeButtonPane</a></li><li><a href="HomeButtonText.html">HomeButtonText</a></li><li><a href="HomeScreen.html">HomeScreen</a></li><li><a href="InnerGridPane.html">InnerGridPane</a></li><li><a href="ListOfArticlesScreen.html">ListOfArticlesScreen</a></li><li><a href="MasterAppBar.html">MasterAppBar</a></li><li><a href="MasterAvatar.html">MasterAvatar</a></li><li><a href="MenuButton.html">MenuButton</a></li><li><a href="MessageDialog.html">MessageDialog</a></li><li><a href="PickMyBrainDialog.html">PickMyBrainDialog</a></li><li><a href="PickMyBrainMessageService.html">PickMyBrainMessageService</a></li><li><a href="PickMyBrainScreen.html">PickMyBrainScreen</a></li><li><a href="SeparatorPane.html">SeparatorPane</a></li><li><a href="SocialCard.html">SocialCard</a></li><li><a href="SocialScreen.html">SocialScreen</a></li><li><a href="TopicScreen.html">TopicScreen</a></li><li><a href="TopicService.html">TopicService</a></li></ul><h3>Global</h3><ul><li><a href="global.html#theme">theme</a></li><li><a href="global.html#title_to_slug">title_to_slug</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue Aug 21 2018 07:30:21 GMT+0000 (UTC)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>