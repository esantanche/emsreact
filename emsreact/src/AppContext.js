// This is the application context
// We populate it in App.js and provide it to the entire application

import React from "react";

const AppContext = React.createContext(null);

export default AppContext;