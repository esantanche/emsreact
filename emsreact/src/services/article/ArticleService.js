// App configuration
import { APP_CONFIGURATION } from '../../appConfiguration';

// Helper function to calculate the number of the next page to fetch
// The first page is page zero
import { number_of_page_to_fetch } from '../../helpers/cache_computations';

/**
 * This service fetches articles from the backend.
 */
class ArticleService {

    constructor() {

        this.articles_cache = {};
        this.articles_count = {};

    }

    /**
     * Function getting articles.
     *
     * There are three implemented filters.
     *
     * The first one fetches a single article. It needs the node id the backend assigned to the article.
     *
     * The second fetches all articles for a given topic. Topics can be "How I work", "The tools I use", etc.
     *
     * The third fetches sticky articles only or all af them.
     *
     * It's possible to combine the topic and sticky filters. Of course the nid filter (single article) has to be used alone.
     *
     * @param {object} filters Filters to use when fetching articles
     * @param {function} callback_to_return_articles_and_more_flag Function to call to return the fetched articles
     */
    get_articles(filters, callback_to_return_articles_and_more_flag) {

        let filter_query_string = "";

        if (filters) {

            if (filters.nid)
                filter_query_string += "&nid=" + filters.nid;

            if (filters.topic)
                filter_query_string += "&field_ems_topic_target_id=" + filters.topic;

            if (filters.sticky)
                filter_query_string += "&sticky=1";

        }

        this.fetch_articles(filter_query_string, (articles_fetched_successfully) => {

            if (articles_fetched_successfully) {

                // I need to know if there are more articles to fetch so that I can return
                // this piece of information and the caller can show the LOAD MORE button

                const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);

                const count_of_articles = this.count_of_articles_for_query(filter_query_string);

                const there_are_more_articles_to_fetch = (count_of_articles > count_of_articles_in_cache);

                const result_to_return = {

                    more: there_are_more_articles_to_fetch,

                    articles: this.articles_cache[filter_query_string]

                };

                callback_to_return_articles_and_more_flag(result_to_return);

            }

        })

    };

    /**
     * Doing the actual fetch. This function will add to the cache an additional page of results or it will do nothing
     * if all articles have already been fetched.
     * If there is nothing in cache, it will fetch page zero.
     * This function returns just true on success and false on failure. The caller will find the articles in cache.
     *
     * @param {string} filter_query_string String with query parameters that will filter the result
     * @param {function} callback Function to call to return true if everything went well, false otherwise.
     */
    fetch_articles(filter_query_string, callback) {

        let page_number;

        // There is a separated cache for each query string
        const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);
        const count_of_articles = this.count_of_articles_for_query(filter_query_string);

        // We need to fetch another page only if we didn't yet fetch all available articles for the
        // given query.
        // If the cache is empty certainly we need to fetch the first page
        if (count_of_articles > count_of_articles_in_cache || count_of_articles_in_cache === 0) {

            if (count_of_articles === -1)
                page_number = 0;
            else
                page_number = number_of_page_to_fetch(count_of_articles_in_cache, count_of_articles, APP_CONFIGURATION.fetchPageSize);

            fetch(APP_CONFIGURATION.backendUrl + "/rest/EMS/view/articlesplayground?_format=json&langcode=en" +
                        filter_query_string + "&page=" + page_number, {
                method: 'GET',
            })
                .then((response) => {
                    if (!response.ok) { throw response }
                    return response.json()
                })
                .then((response) => {

                    this.add_results_to_cache(filter_query_string, response.results);

                    this.update_count_of_articles_for_query(filter_query_string, parseInt(response.count));

                    callback(true);
                })
                .catch((error) => {

                    callback(false);
                });


        } else {

            // Everything is already in cache. The caller has just to read it.

            callback(true);

        }

    };

    /**
     * Returning the number of articles we already have in cache for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Number of articles in cache for the given query string
     */
    count_of_articles_cache_for_query(filter_query_string) {

        // Testing filter_query_string for empty
        // See https://repl.it/@EmanueleSantanc/Check-for-empty-parameter

        if (!filter_query_string) {

            console.error("count_of_articles_cache_for_query, filter_query_string not defined");

            return -1;
        }

        // If the cache for the given query string is not even defined,
        // we define it and set it as empty
        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        return this.articles_cache[filter_query_string].length;

    };

    /**
     * Returning the total number of articles that are available for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Total number of articles that are available for the given
     * query string.
     */
    count_of_articles_for_query(filter_query_string) {

        if (!filter_query_string) {

            console.error("count_of_articles_for_query, filter_query_string not defined");

            return -1;
        }

        // If the total number of articles available for a given query has never been
        // determined, we define it and set it to -1
        if (typeof this.articles_count[filter_query_string] === "undefined") {

            this.articles_count[filter_query_string] = -1;

        }

        return this.articles_count[filter_query_string];

    };

    /**
     * We call this function to update the total number of articles available for a given query.
     * We typically do this after we make an actual query.
     *
     * @param {string} filter_query_string Query string
     * @param {number} count_of_articles Total number of articles that are available for the given
     * query string.
     */
    update_count_of_articles_for_query(filter_query_string, count_of_articles) {

        if (!filter_query_string) {

            console.error("update_count_of_articles_for_query, filter_query_string not defined");

            return;
        }

        this.articles_count[filter_query_string] = count_of_articles;

    };

    /**
     * Adding articles to the cache of articles for the given query string
     *
     * @param {string} filter_query_string Query string
     * @param {array} results Array of additional results to append to the cache
     */
    add_results_to_cache(filter_query_string, results) {

        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        // Append array to array
        // See https://repl.it/@EmanueleSantanc/Append-array-to-an-array

        this.articles_cache[filter_query_string].push(...results);

    }

}

export default ArticleService;