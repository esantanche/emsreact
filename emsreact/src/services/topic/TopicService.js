import { Service } from 'react-services-injector';

// Material UI colors
import red from "@material-ui/core/colors/red";
import lightBlue from "@material-ui/core/colors/lightBlue";
import orange from '@material-ui/core/colors/orange';
import lightGreen from '@material-ui/core/colors/lightGreen';
import yellow from '@material-ui/core/colors/yellow';
import indigo from "@material-ui/core/es/colors/indigo";

/**
 * When showing lists of articles about topics, many details are needed.
 *
 * This service provides these details.
 *
 * Each topic has the following details:
 * * topic name
 * * topic name as it has to be shown in urls
 * * subheader
 * * letter to be used in the avatar
 * * color of avatar
 * * text to be used on the button that links to more articles for the same topic
 */
class TopicService extends Service {

    constructor() {
        super();

        this.topics = {};

        // For now supporting English only
        this.topics['en'] = [

            {
                name: "How I work",
                name_for_url: "how-i-work",
                subheader: "There is value in doing things properly first time",
                letter_for_avatar: "W",
                color_for_avatar: red[500],
                more_articles_button_label: "More about how I work"
            },
            {
                name: "The tools I use",
                name_for_url: "the-tools-i-use",
                subheader: "Modern technology making writing code more productive, fun and easy",
                letter_for_avatar: "T",
                color_for_avatar: lightBlue[400],
                more_articles_button_label: "More about the tools I use"
            },
            {
                name: "Success stories",
                name_for_url: "success-stories",
                subheader: "From Intel Corporation to video streaming of classical music",
                letter_for_avatar: "S",
                color_for_avatar: orange[400],
                more_articles_button_label: "More success stories"
            },
            {
                name: "Testimonials",
                name_for_url: "testimonials",
                subheader: "Happy clients and happy managers",
                letter_for_avatar: "T",
                color_for_avatar: lightGreen[500],
                more_articles_button_label: "More testimonials"
            },
            {
                name: "Talking about my experiences",
                name_for_url: "talking-about-my-experiences",
                subheader: "Sharing my knowledge with the world",
                letter_for_avatar: "E",
                color_for_avatar: yellow[600],
                more_articles_button_label: "More knowledge"
            },
            {
                name: "Where? How?",
                name_for_url: "where-how",
                subheader: "Where do I work? How do I work with you?",
                letter_for_avatar: "W",
                color_for_avatar: indigo[300],
                more_articles_button_label: "More about where and how I work with you"
            }
        ];
    }

    /**
     * Returning all details for all topics.
     *
     * @param {function} callback_to_return_topics Function to call to return all details for all topics
     */
    get_topics(callback_to_return_topics) {

        callback_to_return_topics(this.topics['en']);

    }

    /**
     * Returning all details for a specific topic. The topic name as it appears in urls is given.
     *
     * @param {string} topic_from_url The topic's name as it appears in urls
     * @param {function} callback_to_return_topic Function to call to return the topic's details
     */
    get_topic_from_url(topic_from_url, callback_to_return_topic) {

        const found_topic_details = this.topics['en'].find(topic => topic.name_for_url === topic_from_url);

        callback_to_return_topic(found_topic_details);

    }

}

// "publicName" property is important if you use any kind of minimization on your JS
TopicService.publicName = 'TopicService';

export default TopicService;