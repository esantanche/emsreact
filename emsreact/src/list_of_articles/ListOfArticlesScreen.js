import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Material UI components
import Card from '@material-ui/core/Card';
import Button from "@material-ui/core/Button/Button";

// From components library
import InnerGridPane from '../components_library/panes/InnerGridPane';
import ArticleCard from '../components_library/cards/ArticleCard';

// To provide this component with the application context
// See AppContext.js
import AppContext from "../AppContext";
import SeparatorPane from "../components_library/panes/SeparatorPane";
import CardActions from "@material-ui/core/CardActions";

/**
 * This Screen shows a list of articles. It's supposed to use the service ArticleService
 * to retrieve them from the backend.
 *
 * If the prop sticky is true, only the articles that are flagged as sticky on the backend
 * will be retrieved.
 *
 * @param {string} topic The topic of articles to show. Like "How I work", "The tools I use",
 * "Success stories", etc.
 * @param {boolean} sticky True if only sticky articles should be displayed
 */
class ListOfArticlesScreen extends Component {

    constructor() {
        super();

        this.state = { articles: [], more: undefined }

    }

    static propTypes = {
        topic: PropTypes.string,
        sticky: PropTypes.bool
    };

    componentDidMount() {

        // Fetching articles when the component mounts

        const { ArticleService } = this.props.services;

        this.ArticleService = ArticleService;

        this.get_articles_from_article_service();

    }

    componentDidUpdate(prevProps) {

        // When the topic changes, the props to this component change as well
        // We need to fetch the articles related to the new topic

        // Same thing if the prop 'sticky' changes. Again we have to fetch articles again

        if (prevProps.topic !== this.props.topic ||
            prevProps.sticky !== this.props.sticky) {

            this.get_articles_from_article_service();

        }

    }

    /**
     * Fetching more articles when the user clicks on "LOAD MORE"
     *
     */
    handleLoadMoreButtonClick = () => {

        this.get_articles_from_article_service();
    };

    /**
     * Function that calls the article service to fetch more articles or just the first time
     * to fetch the first page
     */
    get_articles_from_article_service() {

        this.ArticleService.get_articles({ topic: this.props.topic, sticky: this.props.sticky },
            (response) => {

                this.setState({ articles: response.articles, more: response.more });

            }
        )

    }

    render() {

        // We get articles in an array where each item contains one article
        // We produce another array where each item contains two articles
        // This makes it easier to use the component InnerGridPane to build the list.
        // That component shows two articles at a time (a leftComponent and a
        // rightComponent)

        const articles_pairs = this.state.articles.reduce((articles_pairs, article, index, articles) => {

            if (index % 2 === 0)
                articles_pairs.push(articles.slice(index, index + 2));

            return articles_pairs;

        }, []);

        const jsx_for_list_of_articles = articles_pairs.map((pair_of_articles, index) => {

            return (

                <InnerGridPane key={index}

                               leftComponent={(

                                   <ArticleCard image={pair_of_articles[0].field_image}
                                                title={pair_of_articles[0].title}
                                                text_content={pair_of_articles[0].body}
                                                article_node_id={Number(pair_of_articles[0].nid)}>

                                   </ArticleCard>

                               )}

                               rightComponent={pair_of_articles.length === 2 ? (

                                   <ArticleCard image={pair_of_articles[1].field_image}
                                                title={pair_of_articles[1].title}
                                                text_content={pair_of_articles[1].body}
                                                article_node_id={Number(pair_of_articles[1].nid)}>

                                   </ArticleCard>

                               ) : (

                                   <Card></Card>

                               )}

                />

            )

        });

        let more_button;

        // Only if there are more articles to fetch we show the LOAD MORE button
        // If we are fetching sticky articles, we fetch only the first page because
        // we don't want more than one page of articles to be sticky.
        // Sticky articles are what you see on the home page
        if (this.state.more && this.props.sticky !== true) {

            more_button = ( <React.Fragment>
                               <SeparatorPane />
                                <CardActions>
                                    <Button size="small" onClick={this.handleLoadMoreButtonClick.bind(this)}>LOAD MORE</Button>
                                </CardActions>
                            </React.Fragment> );

        }

        return [ jsx_for_list_of_articles, more_button ];

    }

}

export default props => (
    <AppContext.Consumer>
        {services => <ListOfArticlesScreen {...props} services={services} />}
    </AppContext.Consumer>
);
