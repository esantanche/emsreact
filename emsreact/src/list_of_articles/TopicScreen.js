import React, { Component } from 'react';

// Material UI components
import Card from '@material-ui/core/Card';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import CardHeader from '@material-ui/core/CardHeader';

// Material UI icons
import Clear from '@material-ui/icons/Clear';

// From components library
import FrontGridPane from '../components_library/panes/FrontGridPane';
import SeparatorPane from "../components_library/panes/SeparatorPane";

// Screen that shows a list of articles
import ListOfArticlesScreen from '../list_of_articles/ListOfArticlesScreen';

// The injector allows this component to use services, see README.md in folder 'services'
//import { injector } from 'react-services-injector';

import AppContext from '../AppContext';


/**
 * This screen shows a list of articles for a given topic.
 *
 * Both sticky and non-sticky articles are shown.
 *
 * For now there is no pagination. All articles are shown.
 *
 * When this screen is mounted, the url is like this:
 * "/articles/:topic"
 *
 * Props will provide the parameter 'topic' in this variable: this.props.match.params.topic
 */
class TopicScreen extends Component {

    constructor(props) {
        super(props);

        this.state = { topic_details: {} };
    }

    componentDidMount() {

        if (!this.props.match.params.topic) {

            console.error("TopicScreen::componentDidMount");
            console.error("this.props.match.params.topic is not defined");
            console.error("This is a bug");

            return;
        }

        const { TopicService } = this.props.services;

        this.TopicService = TopicService;

        var self = this;

        // First time the component is mounted we find the topic we have to pass to
        // the component ListOfArticlesScreen for it to call the service that will
        // return the articles

        // this.props.match.params.topic is the topic in the format that goes on the url
        // this.state.topic_details.name is the same topic in the format to be used to fetch
        // articles
        // For example:
        // how-i-work is good for urls
        // "How I work" is to be used to fetch articles
        // The service TopicService provides topics good to fetch articles from topics
        // good for urls

        // Consider that the object topic_details returned by get_topic_from_url is an object
        // containing many properties, not just the name of the topic as it is needed
        // to fetch articles

        this.TopicService.get_topic_from_url(this.props.match.params.topic, function(topic_details) {

            self.setState({ topic_details: topic_details });

        });

    }

    componentDidUpdate(prevProps) {

        var self = this;

        // When the topic changes, the props to this component change as well
        // We need to transform again the topic as provided by the url into
        // the topic as needed to fetch articles

        if (prevProps.match.params.topic !== this.props.match.params.topic) {

            if (this.TopicService) {
                this.TopicService.get_topic_from_url(this.props.match.params.topic, function(topic_details) {

                    self.setState({ topic_details: topic_details });

                });
            }

        }

    }

    render() {

        const topic_details = this.state.topic_details;

        // If topic_details.name is not defined, it means that the topic_details object that the service TopicService
        // should provide is not yet available.
        // In this case we just render an empty element

        return (

            topic_details.name ?

            <FrontGridPane>

                <Card>
                    <CardHeader
                        avatar={
                            <Avatar aria-label={topic_details.name} style={{ backgroundColor: topic_details.color_for_avatar }}>
                                {topic_details.letter_for_avatar}
                            </Avatar>
                        }
                        action={
                            <IconButton href="/">
                                <Clear />
                            </IconButton>
                        }
                        title={topic_details.name}
                        subheader={topic_details.subheader}
                    />

                    <ListOfArticlesScreen topic={topic_details.name} sticky={false}/>

                    <SeparatorPane/>

                </Card>

            </FrontGridPane>

                :

            <FrontGridPane>

            </FrontGridPane>

        )

    }

}


export default props => (
    <AppContext.Consumer>
        {services => <TopicScreen {...props} services={services} />}
    </AppContext.Consumer>
);

// export default injector.connect(TopicScreen);

