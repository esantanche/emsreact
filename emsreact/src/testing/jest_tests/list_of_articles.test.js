// Here we test components in the folder list_of_articles

import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import ListOfArticlesScreen from "../../list_of_articles/ListOfArticlesScreen";
import TopicScreen from "../../list_of_articles/TopicScreen";

import ArticleService from "../../services/article/ArticleService";
import TopicService from "../../services/topic/TopicService";

// To provide this component with the application context
// See AppContext.js
import AppContext from "../../AppContext";

const AppArticleService = new ArticleService();
const AppTopicService = new TopicService();

const Services = { TopicService: AppTopicService,
                   ArticleService: AppArticleService,
                   PickMyBrainMessageService: null };

const list_of_articles_screen_to_be_tested = ( <AppContext.Provider value={Services}>
                                                    <ListOfArticlesScreen topic="How I work"
                                                                          sticky={true}
                                                    />
                                               </AppContext.Provider> );

const json_response_rest_api_call = [ {
        "nid": "8",
        "title": "How do I protect my investment in bespoke software?",
        "body": "body of How do I protect my investment in bespoke software?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How_do_I_protect_my_investment_in_bespoke_software.jpg"
    },
    {
        "nid": "7",
        "title": "How do I create my application?",
        "body": "body of How do I create my application?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How%20do%20I%20create%20my%20application%3F.jpg"
    },
    {
        "nid": "6",
        "title": "How do I make my application easy to manage in production?",
        "body": "body of How do I make my application easy to manage in production?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How_do_I_make_my_application_easy_to_manage_in_production.jpg"
    } ];


const json_response_rest_api_call_including_non_sticky = [ {
        "nid": "8",
        "title": "How do I protect my investment in bespoke software?",
        "body": "body of How do I protect my investment in bespoke software?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How_do_I_protect_my_investment_in_bespoke_software.jpg"
    },
    {
        "nid": "7",
        "title": "How do I create my application?",
        "body": "body of How do I create my application?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How%20do%20I%20create%20my%20application%3F.jpg"
    },
    {
        "nid": "6",
        "title": "How do I make my application easy to manage in production?",
        "body": "body of How do I make my application easy to manage in production?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "On",
        "field_image": "\/sites\/default\/files\/2018-07\/How_do_I_make_my_application_easy_to_manage_in_production.jpg"
    },
    {
        "nid": "13",
        "title": "How do I make my application easy to troubleshoot?",
        "body": "body of How do I make my application easy to troubleshoot?",
        "langcode": "English",
        "field_ems_topic": "How I work",
        "sticky": "Off",
        "field_image": "\/sites\/default\/files\/2018-07\/5267178-xsmall.jpg"
    }
];

jest.mock('react-dotdotdot', () => 'div');

describe('Testing ListOfArticlesScreen', function() {

    it('ListOfArticlesScreen renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(list_of_articles_screen_to_be_tested , div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('ListOfArticlesScreen mocking fetch call', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call));

        const div = document.createElement('div');
        ReactDOM.render(list_of_articles_screen_to_be_tested , div);
        ReactDOM.unmountComponentAtNode(div);

        // There has to be one fetch call only
        expect(fetch.mock.calls.length).toEqual(1);

        // The first argument in the fetch call has to be a string
        // containing rest/EMS/view/articles
        // The first argument is the rest api url
        expect(fetch.mock.calls[0][0]).toMatch(/rest\/EMS\/view\/articles/);

    });

    // This test is supposed to return an empty snapshot because the
    // component ListOfArticlesScreen will be rendered before it updates with
    // the response of the fetch
    it('ListOfArticlesScreen renders nothing before receiving fetch response', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call));

        const tree = renderer
            .create(list_of_articles_screen_to_be_tested)
            .toJSON();
        expect(tree).toMatchSnapshot();

    });

});

const topic_screen_to_be_tested = ( <AppContext.Provider value={Services}>
                                        <TopicScreen topic="How I work"
                                                     sticky={true}
                                                     match={{ params: { topic: "how-i-work" }}}
                                        />
                                    </AppContext.Provider> );

describe('Testing TopicScreen', function() {

    it('TopicScreen renders without crashing', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call_including_non_sticky));

        const div = document.createElement('div');
        ReactDOM.render(topic_screen_to_be_tested , div);
        ReactDOM.unmountComponentAtNode(div);

    });

    it('TopicScreen mocking fetch call', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call_including_non_sticky));

        const div = document.createElement('div');
        ReactDOM.render(topic_screen_to_be_tested , div);
        ReactDOM.unmountComponentAtNode(div);

        // There has to be one fetch call only
        expect(fetch.mock.calls.length).toEqual(1);

        // The first argument in the fetch call has to be a string
        // containing rest/EMS/view/articles
        // The first argument is the rest api url
        expect(fetch.mock.calls[0][0]).toMatch(/rest\/EMS\/view\/articles/);

    });

    // This test is supposed to return an almost snapshot because the
    // component TopicScreen will be rendered before it updates with
    // the response of the fetch
    it('TopicScreen renders nothing before receiving fetch response', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call_including_non_sticky));

        const tree = renderer
            .create(topic_screen_to_be_tested)
            .toJSON();
        expect(tree).toMatchSnapshot();

    });

});
