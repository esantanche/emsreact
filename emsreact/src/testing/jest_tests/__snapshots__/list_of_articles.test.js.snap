// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ListOfArticlesScreen ListOfArticlesScreen renders nothing before receiving fetch response 1`] = `null`;

exports[`Testing TopicScreen TopicScreen renders nothing before receiving fetch response 1`] = `
<div
  className="MuiGrid-container-363 MuiGrid-spacing-xs-16-385"
>
  <div
    className="MuiGrid-item-364 MuiGrid-grid-md-1-419 MuiGrid-grid-lg-2-434"
  />
  <div
    className="MuiGrid-item-364 MuiGrid-grid-xs-12-402 MuiGrid-grid-md-10-428 MuiGrid-grid-lg-8-440"
  >
    <div
      className="MuiPaper-root-460 MuiPaper-elevation1-463 MuiPaper-rounded-461 MuiCard-root-459"
    >
      <div
        className="MuiCardHeader-root-487"
      >
        <div
          className="MuiCardHeader-avatar-488"
        >
          <div
            aria-label="How I work"
            className="MuiAvatar-root-493 MuiAvatar-colorDefault-494"
            style={
              Object {
                "backgroundColor": "#f44336",
              }
            }
          >
            W
          </div>
        </div>
        <div
          className="MuiCardHeader-content-490"
        >
          <span
            className="MuiTypography-root-496 MuiTypography-body2-504 MuiCardHeader-title-491"
          >
            How I work
          </span>
          <span
            className="MuiTypography-root-496 MuiTypography-body2-504 MuiTypography-colorTextSecondary-518 MuiCardHeader-subheader-492"
          >
            There is value in doing things properly first time
          </span>
        </div>
        <div
          className="MuiCardHeader-action-489"
        >
          <a
            className="MuiButtonBase-root-526 MuiIconButton-root-520"
            href="/"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
          >
            <span
              className="MuiIconButton-label-525"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-529"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g>
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </g>
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-536"
            />
          </a>
        </div>
      </div>
      <div
        className="SeparatorPane-separatorPane-543"
      >
        Â 
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-364 MuiGrid-grid-md-1-419 MuiGrid-grid-lg-2-434"
  />
</div>
`;
